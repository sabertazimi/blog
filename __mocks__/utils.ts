let navigatorClipboardMocked = false
const writeTextMock = jest.fn()

const mockNavigatorClipboard = () => {
  if (!navigatorClipboardMocked) {
    Object.defineProperty(global.navigator, 'clipboard', {
      value: {
        writeText: writeTextMock,
      },
    })

    navigatorClipboardMocked = true
  }

  return { writeTextMock }
}

export { mockNavigatorClipboard }
