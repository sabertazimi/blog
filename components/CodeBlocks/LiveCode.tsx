import type { Language } from 'prism-react-renderer'
import React from 'react'
import {
  LiveEditor,
  LiveError,
  LivePreview,
  LiveProvider,
} from 'react-live-runner'
import styles from './LiveCode.module.css'
import theme from './monokai'

interface Props {
  language?: Language
  children?: string
}

const scope = {
  ...React,
}

const LiveCode = ({ language, children }: Props): JSX.Element => (
  <div className={styles.container}>
    <LiveProvider
      scope={scope}
      code={children}
      language={language}
      theme={theme}
    >
      <LiveEditor padding="1.25rem" className={styles.editor} />
      <LivePreview className={styles.preview} />
      <LiveError className={styles.error} />
    </LiveProvider>
  </div>
)

export default LiveCode
