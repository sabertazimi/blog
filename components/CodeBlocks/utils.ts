import parseNumericRange from 'parse-numeric-range'
import type { Language } from 'prism-react-renderer'

const normalizeCode = (code: string = '') => code.replace(/\n+$/, '')

function isLanguage(language: string): language is Language {
  return language.length > 0
}

function normalizeLanguage(languageClass?: string): Language {
  if (languageClass != null) {
    const language = languageClass.replace('language-', '')

    if (isLanguage(language))
      return language
  }

  return 'typescript'
}

function normalizeLanguageName(language: string): string {
  switch (language) {
    case 'html':
      return 'HTML'
    case 'xml':
      return 'XML'
    case 'yml':
    case 'yaml':
      return 'YAML'
    case 'css':
      return 'CSS'
    case 'json':
      return 'JSON'
    case 'md':
    case 'markdown':
      return 'Markdown'
    case 'js':
    case 'javascript':
      return 'JavaScript'
    case 'ts':
    case 'typescript':
      return 'TypeScript'
    case 'coffee':
    case 'coffeescript':
      return 'CoffeeScript'
    case 'jsx':
    case 'tsx':
      return 'React'
    case 'objc':
    case 'objectivec':
      return 'Objective-C'
    default:
      return language.charAt(0).toUpperCase() + language.slice(1)
  }
}

function normalizeLines(expression: string): Set<number> {
  return new Set(parseNumericRange(expression))
}

export {
  normalizeCode,
  normalizeLanguage,
  normalizeLanguageName,
  normalizeLines,
}
