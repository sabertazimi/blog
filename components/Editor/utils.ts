function languageToFilepath(language?: string): string {
  switch (language) {
    case 'css':
      return '/styles.css'
    case 'js':
    case 'javascript':
      return '/App.js'
    case 'ts':
    case 'typescript':
      return '/App.ts'
    case 'jsx':
      return '/App.jsx'
    case 'tsx':
      return '/App.tsx'
    case 'vue':
      return '/src/App.vue'
    case undefined:
    default:
      return '/App.tsx'
  }
}

function normalizeFilepath(filename?: string, language?: string): string {
  if (filename == null || filename === '')
    return languageToFilepath(language)

  if (!filename.startsWith('/'))
    return `/${filename}`

  return filename
}

export { normalizeFilepath }
