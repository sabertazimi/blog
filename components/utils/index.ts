import classNames from 'classnames';

export { classNames };
