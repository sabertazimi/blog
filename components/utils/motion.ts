export { AnimatePresence, motion } from 'framer-motion';
export type { MotionProps } from 'framer-motion';
