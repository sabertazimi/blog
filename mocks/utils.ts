let navigatorClipboardMocked = false
const writeTextMock = jest.fn()

function mockNavigatorClipboard() {
  if (!navigatorClipboardMocked) {
    Object.defineProperty(globalThis.navigator, 'clipboard', {
      value: {
        writeText: writeTextMock,
      },
    })

    navigatorClipboardMocked = true
  }

  return { writeTextMock }
}

export { mockNavigatorClipboard }
