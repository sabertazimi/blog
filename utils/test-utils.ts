import { render as testingRender } from '@testing-library/react'
import { ConfigProvider } from 'antd'
import { createElement } from 'react'

export * from '@testing-library/react'

export function render(ui: Parameters<typeof testingRender>[0], options?: Parameters<typeof testingRender>[1]): ReturnType<typeof testingRender> {
  return testingRender(
    createElement(
      ConfigProvider,
      {
        theme: {
          hashed: false,
        },
      },
      ui,
    ),
    options,
  )
}
