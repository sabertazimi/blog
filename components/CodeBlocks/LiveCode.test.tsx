import { render } from '@/utils'
import LiveCode from './LiveCode'

describe('LiveCode', () => {
  it('should render live code correctly (snapshot)', () => {
    const { container } = render(
      <LiveCode language="typescript">const foo = bar();</LiveCode>,
    )

    expect(container).toMatchSnapshot()
  })
})
