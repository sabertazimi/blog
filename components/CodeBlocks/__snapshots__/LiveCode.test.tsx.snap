// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LiveCode should render live code correctly (snapshot) 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="editor"
      style="position: relative; text-align: left; box-sizing: border-box; padding: 0px; overflow: hidden; color: rgb(118, 217, 230); background-color: rgb(42, 42, 42);"
    >
      <pre
        aria-hidden="true"
        style="margin: 0px; border: 0px; background: none; box-sizing: inherit; display: inherit; font-family: inherit; font-style: inherit; font-variant-ligatures: inherit; font-weight: inherit; letter-spacing: inherit; line-height: inherit; tab-size: inherit; text-indent: inherit; text-rendering: inherit; text-transform: inherit; white-space: pre-wrap; word-break: keep-all; overflow-wrap: break-word; position: relative; pointer-events: none; padding: 1.25rem 1.25rem 1.25rem 1.25rem;"
      >
        <span
          class="token-line"
          style="color: rgb(118, 217, 230);"
        >
          <span
            class="token keyword"
            style="color: rgb(249, 38, 114);"
          >
            const
          </span>
          <span
            class="token plain"
          >
             foo 
          </span>
          <span
            class="token operator"
            style="color: rgb(167, 122, 254);"
          >
            =
          </span>
          <span
            class="token plain"
          >
             
          </span>
          <span
            class="token function"
            style="color: rgb(166, 226, 45);"
          >
            bar
          </span>
          <span
            class="token punctuation"
            style="color: rgb(248, 248, 242);"
          >
            (
          </span>
          <span
            class="token punctuation"
            style="color: rgb(248, 248, 242);"
          >
            )
          </span>
          <span
            class="token punctuation"
            style="color: rgb(248, 248, 242);"
          >
            ;
          </span>
        </span>
        

      </pre>
      <textarea
        autocapitalize="off"
        autocomplete="off"
        autocorrect="off"
        class="npm__react-simple-code-editor__textarea"
        data-gramm="false"
        spellcheck="false"
        style="margin: 0px; border: 0px; background: none; box-sizing: inherit; display: inherit; font-family: inherit; font-style: inherit; font-variant-ligatures: inherit; font-weight: inherit; letter-spacing: inherit; line-height: inherit; tab-size: inherit; text-indent: inherit; text-rendering: inherit; text-transform: inherit; white-space: pre-wrap; word-break: keep-all; overflow-wrap: break-word; position: absolute; top: 0px; left: 0px; height: 100%; width: 100%; resize: none; overflow: hidden; padding: 1.25rem 1.25rem 1.25rem 1.25rem;"
      >
        const foo = bar();
      </textarea>
      <style>
        
/**
 * Reset the text fill color so that placeholder is visible
 */
.npm__react-simple-code-editor__textarea:empty {
  -webkit-text-fill-color: inherit !important;
}

/**
 * Hack to apply on some CSS on IE10 and IE11
 */
@media all and (-ms-high-contrast: none), (-ms-high-contrast: active) {
  /**
    * IE doesn't support '-webkit-text-fill-color'
    * So we use 'color: transparent' to make the text transparent on IE
    * Unlike other browsers, it doesn't affect caret color in IE
    */
  .npm__react-simple-code-editor__textarea {
    color: transparent !important;
  }

  .npm__react-simple-code-editor__textarea::selection {
    background-color: #accef7 !important;
    color: transparent !important;
  }
}

      </style>
    </div>
    <div
      class="preview"
    />
    <pre
      class="error"
    >
      ReferenceError: bar is not defined
    </pre>
  </div>
</div>
`;
